### PR TITLE
Cleanup of CloudSQL logic and fixup LastOperation

### PR DESCRIPTION
**Pull request by josephlewis42**
This commit cleans up the way CloudSQL was halding operations.
Rather than keeping a separate table for operations, they now get stored
with a state tag and operation name as part of the instance information.
This allows us to comply with the OSB spec better and approach things
more transactionally with fewer failure states.

Storing the state information in the instance with a well-defined
state machine also means we can reason about the state of the service
in the gcp_service_broker implementation without resorting to
separate function calls that had to be implemented on BrokerHelpers.

Finally, it has allowed us to do away with the last special case that
our DAO had to deal with and the entire DAO entry for CloudOperations
which was about a 600 line savings.

### CODE REVIEW

File: brokerapi/brokers/account_managers/service_account_manager.go [hunk 1]
```diff
@@ -133,11 +134,12 @@ func (sam *ServiceAccountManager) DeleteCredentials(ctx context.Context, binding
 		return fmt.Errorf("Error creating IAM service: %s", err)
 	}
 	saService := iam.NewProjectsServiceAccountsService(iamService)
+	resourceName := projectResourcePrefix + sam.ProjectId + "/serviceAccounts/" + saCreds.UniqueId
+	if _, err = saService.Delete(resourceName).Do(); err != nil {
```
Comment (craigatgoogle) [inline brokerapi/brokers/account_managers/service_account_manager.go]: Consider scope restricting the err variable to the if block using: if _, err := saService.Delete(resourceName).Do(); err != nil

This approach is better from a defensive programming stand-point. Also consider the same approach for the if blocks throughout this function.

File: brokerapi/brokers/gcp_service_broker.go [hunk 2]
```diff
@@ -394,13 +392,8 @@ func (gcpBroker *GCPServiceBroker) lastOperationAsync(ctx context.Context, insta
...
 
 	// no error and we're done! Delete from the SB database if this was a delete flow and return success
-	deleteFlow, err := service.LastOperationWasDelete(ctx, instanceId)
-	if err != nil {
-		return brokerapi.LastOperation{State: brokerapi.Succeeded}, fmt.Errorf("Couldn't determine if provision or deprovision flow, this may leave orphaned resources, contact your operator for cleanup: %s", err)
-	}
-	if deleteFlow {
-		err = db_service.DeleteServiceInstanceDetailsById(ctx, instanceId)
-		if err != nil {
+	if lastOperationType == models.DeprovisionOperationType {
```
[the remark this answers is not in the document]
Comment (josephlewis42) [reply brokerapi/brokers/gcp_service_broker.go]: This method isn't the cleanest thing. The idea is that some delete operations in the OSB spec are asynchronous and we shouldn't be removing the actual row until we know it succeeded. The OSB client will--upon receiving a response code indicating that the delete was async--continue to poll lastOperation until the method succeeds or fails. This function gets brought back in and if it was a delete flow cleans it up.

I was thinking about refactoring this broker into an adapter pattern for the OSB spec so each individual broker could act as a standalone service broker, then adding a router over top which would select the correct one for a given request. Thoughts? In the short-term I can pull this out into a function.

**Comment by josephlewis42** [comment]
@craigatgoogle I refactored this so the state management is now (almost fully) in the `gcp_service_broker.go` file. Cleanup on success either clears out the existing operation on the object or deletes the object, depending on where it was transitioning from.

The good news being, except for the CloudSQL broker, brokers no longer import "db_service", which I think is a huge win compared to where we started out. Once CloudSQL and the account management code get refactored, all db work will be done in `gcp_service_broker.go`.

File: brokerapi/brokers/gcp_service_broker.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (craigatgoogle) [reply brokerapi/brokers/gcp_service_broker.go]: Sounds good to me. Thanks.

**Pull request merged**